### PR TITLE
Add support for wrap/unwrap RSA with aes_cbc_pad.

### PR DESCRIPTION
**Pull request by williamrettig**
Sean, David,

Here are the updates for the C_WrapKey/C_UnwrapKey addition that we (BRN) needs.  Please consider adding this.

As discussed on the call, we haven't added unit tests; however, Todd indicated that you would be willing.  We tested by implementation compatibility checks in a larger use case. All possible return codes haven't been tested.

### CODE REVIEW

File: src/crypto.c [hunk 1]
```diff
@@ -3676,21 +3824,82 @@ CK_RV C_WrapKey(CK_SESSION_HANDLE hSession,
...
+            if (ret != 0)
+                return CKR_FUNCTION_FAILED;
+
+            serialBuff = XMALLOC(serialSize, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            if (serialBuff == NULL)
+                return CKR_HOST_MEMORY;
+
+            ret = WP11_Rsa_SerializeKey(key,  serialBuff, &serialSize);
+            if (ret != 0)
+                return CKR_FUNCTION_FAILED;
```
Comment (williamrettig) [inline src/crypto.c:3861]: Memory leak.  Free allocated serialBuf prior to return

Comment (williamrettig) [reply src/crypto.c]: commit 308db0f00271c0f00f82f119dba30cc53877d009 (HEAD -> master, origin/master, origin/HEAD)
Author: williamrettig <william.rettig@outlook.com>
Date:   Wed Aug 24 17:13:00 2022 -0400

    resolve review comment 1, correct potential memory leak

**Comment by dgarske** [comment]
@SparkiDev please review this when you have time. We are getting closer to having the legals worked out. Thanks

File: src/crypto.c [hunk 2]
```diff
@@ -3676,21 +3824,84 @@ CK_RV C_WrapKey(CK_SESSION_HANDLE hSession,
...
+    keyClass = WP11_Object_GetClass(key);
+
+    rv = CHECK_WRAPPABLE(keyClass, keyType);
+    if (rv != CKR_OK)
+        return rv;
+
+    switch(keyType) {
+
+        case CKK_RSA:
+            ret = WP11_Rsa_SerializeKey(key, NULL, &serialSize);
```
Comment (williamrettig) [inline src/crypto.c:3851]: Mere DER serialization is not sufficient for private keys.  Needs to adhere to PKCS8.  Wolfcrypt already supports PKCS8 in asn1 module.  Add function to serialize in PKCS8 after serializing to DER.

Comment (williamrettig) [reply src/crypto.c]: https://github.com/wolfSSL/wolfPKCS11/pull/15/commits/06e6159c73511e23328386802a1a7c32883b40c4

Comment (williamrettig) [reply src/crypto.c]: Updated with PKCS8 plain text version

**Comment by dgarske** [comment]
Hi @williamrettig thanks for these updates. Sean and I will review in detail soon.

**Comment by williamrettig** [comment]
David, Sean,

Was/is a pleasure.  Thank you for the willingness.

I regret that I didn't pay strong attention to the test code.  I note that I changed a few return codes that 'seemed' more appropriate such as differentiating between an object handle and a wrap/unwrap key handle.

Bill
________________________________

**Comment by dgarske** [comment]
Hi @williamrettig ,

That is alright. Sean and I have been quite busy. We will review soon. If you are able to expand the tests for this that would be wonderful. If you'd like to schedule some time to chat please let us know.

Thanks,
David Garske, wolfSSL

**Comment by williamrettig** [comment]
David,

Yes, let's have a chat; I'd like that.

Bill
________________________________

**Review by dgarske — APPROVED** [review]
Contribution approved via section 4 of license agreement and email from customer. Approval from CEO Larry.

**Pull request merged**